etcdserver/api/membership/cluster.go

```diff
@@ -811,23 +859,70 @@ func ValidateClusterAndAssignIDs(lg *zap.Logger, local *RaftCluster, existing *R
 	return nil
 }
 
-func mustDetectDowngrade(lg *zap.Logger, cv *semver.Version) {
+func mustDetectDowngrade(lg *zap.Logger, cv *semver.Version, d *DowngradeInfo) {
 	lv := semver.Must(semver.NewVersion(version.Version))
 	// only keep major.minor version for comparison against cluster version
 	lv = &semver.Version{Major: lv.Major, Minor: lv.Minor}
+
+	// if the cluster enables downgrade, check local version against downgrade target version.
+	if d != nil && d.Enabled && d.TargetVersion != nil {
+		if isValidDowngrade(d.TargetVersion, lv) {
+			if cv != nil {
+				if lg != nil {
+					lg.Info(
+						"cluster is downgrading to target version",
+						zap.String("target-cluster-version", d.TargetVersion.String()),
+						zap.String("determined-cluster-version", version.Cluster(cv.String())),
+						zap.String("current-server-version", version.Version),
+					)
+				}
+			}
+			return
+		}
+		if lg != nil {
+			lg.Fatal(
+				"invalid downgrade; server version is not allowed to join when downgrade is enabled",
+				zap.String("current-server-version", version.Version),
+				zap.String("target-cluster-version", d.TargetVersion.String()),
+			)
+		}
+	}
+
+	// if the cluster disables downgrade, check local version against determined cluster version.
+	// the validation passes when local version is not less than cluster version
 	if cv != nil && lv.LessThan(*cv) {
 		if lg != nil {
 			lg.Fatal(
 				"invalid downgrade; server version is lower than determined cluster version",
 				zap.String("current-server-version", version.Version),
 				zap.String("determined-cluster-version", version.Cluster(cv.String())),
 			)
-		} else {
-			plog.Fatalf("cluster cannot be downgraded (current version: %s is lower than determined cluster version: %s).", version.Version, version.Cluster(cv.String()))
 		}
 	}
 }
 
+// isValidDowngrade checks whether joining local server is a valid downgrade when cluster enables downgrade.
+// the validation passes when local version is one minor version higher then target version
+func isValidDowngrade(tv *semver.Version, lv *semver.Version) bool {
```

Comment:
this needs unit tests?

Reply:
Yes, it has unit test but not in this pr. I thought it will be a big pr if I include all the tests. Should I put them in or just add the unit tests?